usb/manual/ssdt.md

```diff
@@ -0,0 +1,165 @@
+## SSDT method
+
+A little bit of history worth to mention:
+
+* Thanks to [Gengik84](https://www.macos86.it/profile/1-gengik84/) and the [macos86.it community](https://www.macos86.it/topic/9-mappatura-porte-usb/) for developing and testing this quick and easy method, the original post dates back to 2018, even though some folks [stole it later in 2019](https://www.olarila.com/topic/6181-guide-native-usb-fix-for-desktops-no-injectorkext-required-skylake/) without mentioning any credits.
+
+## Prerequisites:
+
+* This guide assumes that the user knows how to use MaciASL, IORegistryExplorer and ProperTree. We'll use a sample motherboard, the ASUS Z370 Prime A II where the USB devices are described in the `SSDT_xh-rvp08.aml`. 
+* Use OpenCore's `SysReport` feature to look where your OEM put the definitions for your USB devices in the ACPI tables (if none found in the SSDTs, your last chance should be your DSDT)
+
+Before any modification, write down the `Length` value which is located in the beginning of the file, as shown: 
+ 
+![](../../images/post-install/usb-md/usb-map-ssdt-table-lenght.png)  
+
+The unmodified, original ACPI table has a lenght of `2835`, this is going to be useful later in the guide, when we're going to tell OpenCore to avoid loading that table in favour of the new one we're going to modify.
```

Comment:
> the ACPI for the root hub is located in the DSDT, so this is going to be useless.

The `RHUB` in the DSDT only defines the `_ADR` of the port not the connector type. This method is valid for skylake and newer (I have to add this warning in the guide) hence it is necessary to block the relative SSDT which uses `GUPC` or similar

Reply:
In many cases the ports are defined later in the DSDT; for example:

<details>
<summary>Skylake laptop</summary>

```
Scope (_SB.PCI0.XHC.RHUB.HS01)
{
    Method (_UPC, 0, Serialized)  // _UPC: USB Port Capabilities
    {
        Return (UUPC (0xFF, 0x03))
    }

    Method (_PLD, 0, Serialized)  // _PLD: Physical Location of Device
    {
        Return (UPLD (One, One))
    }
}

Scope (_SB.PCI0.XHC.RHUB.HS02)
{
    Method (_UPC, 0, Serialized)  // _UPC: USB Port Capabilities
    {
        Return (UUPC (0xFF, 0x03))
    }

    Method (_PLD, 0, Serialized)  // _PLD: Physical Location of Device
    {
        Return (UPLD (One, 0x02))
    }
}
...
```

</details>

<details>
<summary>AMD desktop, PTXH</summary>

```
Device (PTXH)
{
    Name (_ADR, Zero)  // _ADR: Address
    Method (_PRW, 0, NotSerialized)  // _PRW: Power Resources for Wake
    {
        Return (GPRW (0x19, 0x04))
    }

    Device (RHUB)
    {
        Method (GPLD, 2, Serialized)
        {
            Name (PCKG, Package (0x01)
            {
                Buffer (0x10){}
            })
            CreateField (DerefOf (PCKG [Zero]), Zero, 0x07, REV)
            REV = One
            CreateField (DerefOf (PCKG [Zero]), 0x40, One, VISI)
            VISI = Arg0
            CreateField (DerefOf (PCKG [Zero]), 0x57, 0x08, GPOS)
            GPOS = Arg1
            Return (PCKG) /* \_SB_.PCI0.GPP2.PTXH.RHUB.GPLD.PCKG */
        }

        Method (GUPC, 1, Serialized)
        {
            Name (PCKG, Package (0x04)
            {
                Zero, 
                0xFF, 
                Zero, 
                Zero
            })
            PCKG [Zero] = Arg0
            Return (PCKG) /* \_SB_.PCI0.GPP2.PTXH.RHUB.GUPC.PCKG */
        }

        Name (_ADR, Zero)  // _ADR: Address
        Device (SS01)
        {
            Name (_ADR, One)  // _ADR: Address
            Method (_UPC, 0, NotSerialized)  // _UPC: USB Port Capabilities
            {
                Return (GUPC (One))
            }

            Method (_PLD, 0, NotSerialized)  // _PLD: Physical Location of Device
            {
                Return (GPLD (One, 0x06))
            }
        }

        Device (SS02)
        {
            Name (_ADR, 0x02)  // _ADR: Address
            Name (_UPC, Package (0x04)  // _UPC: USB Port Capabilities
            {
                0xFF, 
                0x09, 
                Zero, 
                Zero
            })
            Method (_PLD, 0, NotSerialized)  // _PLD: Physical Location of Device
            {
                Return (GPLD (One, One))
            }
        }
...
```

</details>

<details>
<summary>AMD desktop, XHC0</summary>

```
Scope (_SB.PCI0.GP13.XHC0)
{
    Device (RHUB)
    {
        Name (_ADR, Zero)  // _ADR: Address
        Device (PRT1)
        {
            Name (_ADR, One)  // _ADR: Address
            Name (_UPC, Package (0x04)  // _UPC: USB Port Capabilities
            {
                0xFF, 
                0x03, 
                Zero, 
                Zero
            })
            Name (_PLD, Package (0x01)  // _PLD: Physical Location of Device
            {
                ToPLD (
                    PLD_Revision           = 0x2,
                    PLD_IgnoreColor        = 0x1,
                    PLD_Red                = 0x0,
                    PLD_Green              = 0x0,
                    PLD_Blue               = 0x0,
                    PLD_Width              = 0x0,
                    PLD_Height             = 0x0,
                    PLD_UserVisible        = 0x1,
                    PLD_Dock               = 0x0,
                    PLD_Lid                = 0x0,
                    PLD_Panel              = "BACK",
                    PLD_VerticalPosition   = "CENTER",
                    PLD_HorizontalPosition = "LEFT",
                    PLD_Shape              = "VERTICALRECTANGLE",
                    PLD_GroupOrientation   = 0x0,
                    PLD_GroupToken         = 0x0,
                    PLD_GroupPosition      = 0x1,
                    PLD_Bay                = 0x0,
                    PLD_Ejectable          = 0x1,
                    PLD_EjectRequired      = 0x0,
                    PLD_CabinetNumber      = 0x0,
                    PLD_CardCageNumber     = 0x0,
                    PLD_Reference          = 0x0,
                    PLD_Rotation           = 0x0,
                    PLD_Order              = 0x0,
                    PLD_VerticalOffset     = 0xFFFF,
                    PLD_HorizontalOffset   = 0xFFFF)

            })
        }

        Device (PRT2)
        {
            Name (_ADR, 0x02)  // _ADR: Address
            Name (_UPC, Package (0x04)  // _UPC: USB Port Capabilities
            {
                0xFF, 
                0x03, 
                Zero, 
                Zero
            })
            Name (_PLD, Package (0x01)  // _PLD: Physical Location of Device
            {
                ToPLD (
                    PLD_Revision           = 0x2,
                    PLD_IgnoreColor        = 0x1,
                    PLD_Red                = 0x0,
                    PLD_Green              = 0x0,
                    PLD_Blue               = 0x0,
                    PLD_Width              = 0x0,
                    PLD_Height             = 0x0,
                    PLD_UserVisible        = 0x1,
                    PLD_Dock               = 0x0,
                    PLD_Lid                = 0x0,
                    PLD_Panel              = "BACK",
                    PLD_VerticalPosition   = "CENTER",
                    PLD_HorizontalPosition = "RIGHT",
                    PLD_Shape              = "VERTICALRECTANGLE",
                    PLD_GroupOrientation   = 0x0,
                    PLD_GroupToken         = 0x0,
                    PLD_GroupPosition      = 0x2,
                    PLD_Bay                = 0x0,
                    PLD_Ejectable          = 0x1,
                    PLD_EjectRequired      = 0x0,
                    PLD_CabinetNumber      = 0x0,
                    PLD_CardCageNumber     = 0x0,
                    PLD_Reference          = 0x0,
                    PLD_Rotation           = 0x0,
                    PLD_Order              = 0x0,
                    PLD_VerticalOffset     = 0xFFFF,
                    PLD_HorizontalOffset   = 0xFFFF)

            })
        }
...
```

</details>

Reply:
Again, this is out of the scope of this guide - it is called "SSDT Method" not "DSDT/SSDT method" 

I cannot recommend to inject the DSDT, every other piece of your guide says: "Do not load a custom DSDT, is already in your firmware"

Reply:
Blocking the original OEM SSDT table for the USB works on our end, the _STA is not necessary

AMD is not supported in this guide, nor laptops since you have to edit the DSDT.

Reply:
It is feasible - but every other part of the guide who says "Do not inject your DSDT"

Reply:
> It is feasible - but every other part of the guide who says "Do not inject your DSDT" has to be updated

Or you can `_STA` the old root hub and create a new one in an SSDT, as I mentioned above. This would also avoid having to drop any tables.

Reply:
@dhinakg Hello, I am AMD notebook, there are two buses xhc0 and xhc1, xhc1 must be turned off to enter the system, I saw the above document can customize DSDT, can this method solve the problem of not closing xhc1